File: cpp/1026_MaximumDifferenceBetweenNodeAndAncestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        if (!root) return 0;
        return helper(root, INT_MAX, INT_MIN);
    }
    
    int helper(TreeNode* root, int small, int big) {
        if (!root) return big - small;
        
        small = min(small, root->val), big = max(big, root->val);
        return max(helper(root->left, small, big), helper(root->right, small, big));
    }
};

File: java/1026_MaximumDifferenceBetweenNodeAndAncestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxAncestorDiff(TreeNode root) {
        return findMaxDiff(root, Integer.MAX_VALUE, Integer.MIN_VALUE);
    }
    
    public int findMaxDiff(TreeNode root, int low, int high) {
        if (root == null) return high - low;
        
        high = Math.max(high, root.val);
        low = Math.min(low, root.val);
        
        return Math.max(findMaxDiff(root.left, low, high), findMaxDiff(root.right, low, high));
    }
}
